src/gen_rules.ml

```diff
@@ -618,6 +628,41 @@ module Gen(P : Params) = struct
        Build.store_vfile vrequires);
     Build.vpath vrequires
 
+  let requires_and_dot_merlin ~dir ~dep_kind ~item ~libraries ~preprocess ~virtual_deps ~alias_module =
+    (* Compute requires *)
+    let requires = requires ~dir ~dep_kind ~item ~libraries ~preprocess ~virtual_deps in
+
+    (* Only generate .merlin file in default context *)
+    match Path.extract_build_context dir with
+    | Some ("default", remaindir) ->
+        let path = Path.relative remaindir ".merlin" in
+        add_rule (
+          let dot_merlin = ref [] in
+          let add_line s = dot_merlin := s :: !dot_merlin in
+          add_line "S .";
+          add_line ("B " ^ Path.reach dir ~from:remaindir);
+
+          (* Separate internal and external libraries *)
+          let internals, externals = Lib_db.interpret_lib_deps_for_dot_merlin ~dir libraries in
```

Comment:
You shouldn't need to call this function. `requires` already gives you the list of resolved libraries, it has type `(unit, Lib.t list) Build.t`. So you can just write:

```ocaml
requires
>>^ fun libs ->
List.partition_map libs ~f:(function
  | Internal (path, lib) -> ...
  | External pkg -> ...)
```